mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -16,49 +17,79 @@ use super::EthCoin;
 use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
             BalanceError, MmCoin};
 
+type BalanceResult = Result<(String, HashMap<String, BigDecimal>), (String, MmError<BalanceError>)>;
+
 /// This implementation differs from others, as they immediately return
 /// an error if any of the requests fails. This one completes all futures
 /// and returns their results individually.
-async fn get_all_balance_results_concurrently(
-    coin: &EthCoin,
-) -> Vec<Result<(String, BigDecimal), (String, MmError<BalanceError>)>> {
+async fn get_all_balance_results_concurrently(coin: &EthCoin) -> Vec<BalanceResult> {
     let mut tokens = coin.get_erc_tokens_infos();
 
-    // Workaround for performance purposes.
-    //
-    // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
-    // type and mapping the platform coin and the entire token list (which can grow at any time), we map
-    // the platform coin to Erc20TokenInfo so that we can use the token list right away without
-    // additional mapping.
-    tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
-        token_address: coin.my_address,
-        decimals: coin.decimals,
-    });
-
-    let jobs = tokens
-        .into_iter()
-        .map(|(token_ticker, info)| async move { fetch_balance(coin, token_ticker, &info).await })
-        .collect::<FuturesUnordered<_>>();
-
-    jobs.collect().await
+    let addresses = match coin.my_addresses().await {
+        Ok(addresses) => addresses,
+        Err(e) => {
+            return vec![Err((coin.ticker.clone(), e.into()))];
+        },
+    };
+
+    let mut all_jobs = FuturesUnordered::new();
+
+    for address in addresses {
+        // Workaround for performance purposes.
+        //
+        // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
+        // type and mapping the platform coin and the entire token list (which can grow at any time), we map
+        // the platform coin to Erc20TokenInfo so that we can use the token list right away without
+        // additional mapping.
+        tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
+            token_address: address,
+            decimals: coin.decimals,
+        });
```

Comment:
Don't we have those addresses in the `erc20_tokens_infos` already? If not, I think we should if I am not missing something.

Reply:
I followed what you did there https://github.com/KomodoPlatform/komodo-defi-framework/blob/96fac7c8c97b5c2267bfd9bd250237817269acfe/mm2src/coins/eth/eth_balance_events.rs#L27-L36 but I guess I made a mistake. We only need the platform ticker once, I guess I can use all zeros address for the token address of the platform coin and add it before the loop, will fix this.

Reply:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/37e2b7102152bc6655465b2c6c91df1876c4519f